map/src/main/kotlin/edu/artic/map/MapViewModel.kt

```diff
@@ -0,0 +1,200 @@
+import io.reactivex.Observable
+import io.reactivex.functions.BiFunction
+import io.reactivex.functions.Function3
+import io.reactivex.functions.Function4
+import io.reactivex.subjects.BehaviorSubject
+import io.reactivex.subjects.Subject
+import timber.log.Timber
+import javax.inject.Inject
+
+class MapViewModel @Inject constructor(
+        private val mapAnnotationDao: ArticMapAnnotationDao,
+        private val galleryDao: ArticGalleryDao,
+        private val objectDao: ArticObjectDao
+) : BaseViewModel() {
+
+    val mapAnnotations: Subject<List<MapItem<*>>> = BehaviorSubject.create()
+
+    val alwaysVisibleAnnotations: Subject<List<MapItem.Annotation>> = BehaviorSubject.create()
+
+    val floor: Subject<Int> = BehaviorSubject.createDefault(1)
+    val zoomLevel: Subject<MapZoomLevel> = BehaviorSubject.create()
+
+    val currentFloor: Int
+        get() {
+            val castFloor = (floor as BehaviorSubject)
+            return if (castFloor.hasValue()) {
+                castFloor.value
+            } else {
+                1
+            }
+        }
+
+    val currentZoomLevel: MapZoomLevel
+        get() {
+            val castLevel = (zoomLevel as BehaviorSubject)
+            return if (castLevel.hasValue()) {
+                castLevel.value
+            } else {
+                MapZoomLevel.One
+            }
+        }
+
+    init {
+        val allAmenities = mapAnnotationDao
+                .getAmenitiesOnMap()
+                .map { annotationList ->
+                    val list = mutableListOf<MapItem.Annotation>()
+                    annotationList.forEach { annotation ->
+                        val floor = annotation.floor
+                                .let {
+                                    it?.toInt() ?: 1
+                                }
+                        list.add(MapItem.Annotation(annotation, floor))
+                    }
+                    return@map list
+
+                }
+
+        val allBuildingNames = mapAnnotationDao
+                .getBuildingNamesOnMap()
+                .map { annotationList ->
+                    val list = mutableListOf<MapItem.Annotation>()
+                    annotationList.forEach { annotation ->
+                        val floor = annotation.floor
+                                .let {
+                                    it?.toInt() ?: 1
+                                }
+                        list.add(MapItem.Annotation(annotation, floor))
+                    }
+                    return@map list
+
+                }
+
+        Observable.combineLatest(
+                allAmenities.asObservable(),
+                allBuildingNames.asObservable(),
+                BiFunction<List<MapItem.Annotation>, List<MapItem.Annotation>, List<MapItem.Annotation>>
+                { amenities, buildingNames ->
+                    val list = mutableListOf<MapItem.Annotation>()
+                    list.addAll(amenities)
```

Comment:
idiomatic kotlin you probably could do: `return @BiFunction amenities.toMutableList() + buildingNames` which copies the list over and adds the other list

Reply:
that i did not know thanks for that shortcut

Reply:
now you know :) you can use those `toList()` operations to convert between mutable and immutable. it copies over the list if you look into the source.

Reply:
done